scalameta/contrib/src/main/scala/scala/meta/contrib/DocToken.scala

```diff
@@ -0,0 +1,275 @@
+package scala.meta.contrib
+
+import fastparse.all._
+import fastparse.core.Parsed
+
+/**
+  * Represents a scaladoc line.
+  */
+case class DocToken(kind: DocToken.Kind, name: Option[String], body: Option[String]) {
+
+  override def toString: String = {
+    ((name, body) match {
+      case (Some(n), _) => s"$kind(name=$n, body=${body.getOrElse("")})"
+      case (None, Some(kindBody)) => s"$kind($kindBody)"
+      case _ => kind.toString
+    }).replaceAll("\n", " ")
+  }
+
+  /**
+    * Obtains all the references from a [[DocToken]].
+    */
+  def references: Seq[DocToken.Reference] =
+    body
+      .map { b =>
+        {
+          def parseBodyFrom(idx: Int): Seq[DocToken.Reference] = {
+            DocToken.referenceParser
+              .parse(b, idx) match {
+              case p: Parsed.Success[String, _, _] =>
+                Seq(DocToken.Reference(p.value)) ++ parseBodyFrom(p.index)
+              case _ => Seq[DocToken.Reference]()
+            }
+          }
+
+          parseBodyFrom(0)
+        }
+      }
+      .getOrElse(Nil)
+}
+
+/**
+  * Companion object containing the classes required for describing an ScalaDoc token.
+  *
+  * The available tokens and their documentation are obtained from:
+  * @see http://docs.scala-lang.org/overviews/scaladoc/for-library-authors.html
+  */
+object DocToken {
+
+  /**
+    * Parser that for obtaining a class reference like '[[scala.Some]]' from an scaladoc body.
+    */
+  private val referenceParser = P(
+    // Removes the elements previous to the references.
+    ((AnyChar ~ !"[[").rep ~ AnyChar).?
```

Comment:
I know I was the one to recommend fastparse, but i'd like to mention that regexes are actually perfect for simple stuff like this.

```scala
@ val Reference = "\\[\\[([^]]+)\\]\\]".r
Reference: Regex = \[\[([^]]+)\]\]
@ Reference.findAllMatchIn("This is [[Option]], [banana] is [[cool]].").map(_.group(1)).toList
res36: List[String] = List("Option", "cool")

@ val Codeblock = "(?s)\\{\\{\\{(.*?)\\}\\}\\}".r
Codeblock: Regex = (?s)\{\{\{(.*?)\}\}\}
@ Codeblock.findAllMatchIn(""" Doc
  {{{

    val x = 2
  }}}

  foo

  {{{ val y = 2 }}}

  banana""").map(_.group(1)).toList
res41: List[String] = List(
  """


  val x = 2

  """,
  " val y = 2 "
)
```